backend/layers/business/business.py

```diff
@@ -552,9 +552,23 @@ def get_dataset_version(self, dataset_version_id: DatasetVersionId) -> Optional[
     def get_dataset_version_from_canonical(self, dataset_id: DatasetId) -> Optional[DatasetVersion]:
         """
         Given a canonical dataset id, returns its mapped dataset version, i.e. the dataset version
-        that belongs to the most recently published collection
+        that belongs to the most recently published collection. Otherwise will return the most recent
+        unpublished version.
         """
-        return self.database_provider.get_dataset_mapped_version(dataset_id)
+        dataset_version = self.database_provider.get_dataset_mapped_version(dataset_id)
+        if dataset_version is not None:
+            return dataset_version
+        else:
+            canonical_dataset = self.database_provider.get_canonical_dataset(dataset_id)
+            if not canonical_dataset:
+                return None
+            latest = datetime.datetime.fromtimestamp(0)
```

Comment:
I would add a comment that explains why you do this (i.e. you want to lookup the latest version in case there are multiple versions in an unpublished collection).

Also, you could avoid the loop by doing `unpublished_dataset = max(dataset_versions, key=lambda d: d.created_at`, but I don't know if the functional style is preferred to the iterative style in our codebase.

Reply:
I'm going to stick to iterative for now since we do it in other parts of the codebase (i.e.`get_unpublished_collection_version_from_canonical`) but I don't feel strongly about it

will add the comment though